aip/general/0202.md

```diff
@@ -0,0 +1,104 @@
+information.
+
+## Guidance
+
+Decorating a field with `google.api.field` is only necessary when explicitly
+stated in this AIP or another that leverages `google.api.Field` information. As
+such, the guidance herein applies to those scenarios as well.
+
+### Format
+
+Fields with a primitive type can still have a specific format. To convey that
+type format, the `Field.Format` enumeration is used via the
+`(google.api.field).format` extension field. The following guidance conveys the
+meaning of and requirements for use of each `Field.Format` value.
+
+#### UUID4
+
+The `UUID4` format represents a UUID version 4 value as governed by
+[RFC 4122][]. It **must** only be used on a field of type `string`.
+
+Such a value **may** be normalized by the service to entirely lowercase letters.
+For example, the value `F47AC10B-58CC-0372-8567-0E02B2C3D479` would be
+normalized to `f47ac10b-58cc-0372-8567-0e02b2c3d479`.
+
+#### IPv4
+
+The `IPV4` format represents an IP v4 address as governed by [RFC 791][]. It
+**must** only be used on a field of type `string`.
+
+Such a value **may** be condensed by the service, with leading zeros in each
+octet stripped. For example, `001.022.233.040` would be condensed to
+`1.22.233.40`.
+
+#### IPv6
+
+The `IPV6` format represents an IP v6 address as governed by [RFC 2460][]. It
+**must** only be used on a field of type `string`.
+
+Such a value **may** be normalized by the service to entirely lowercase letters,
+and zero-padded partial and empty octets. For example, the value `2001:DB8::`
+would be normalized to `2001:0db8:0000:0000`.
+
+#### IPv4 or IPv6
+
+The `IPV4_OR_IPV6` value indicates that the field can be either an IP v4 or v6
+address, as described in the [IPv4](#ipv4) and [IPv6](#ipv6) sections.
+
+#### Format Compatibility
+
+Adding a format specifier to an existing, unspecified field **is not** backwards
+compatible, *unless* the field in question has always conformed to the format
+being specified.
+
+Changing an existing format specifier to a different one in all cases **is not**
+backwards compatible.
+
+#### Extending Format
+
+Any new `Field.Format` value **must** be governed by an
+[IETF-approved RFC][ietf rfc] or a [Google-approved AIP](./0001.md).
+
+## Rationale
+
+#### Why add a format specifier?
+
+The format of a primitive-typed field can be critical to its usability. Some
+programming languages may convey a specific type format as a standalone type,
+as Java does with [UUID][java uuid]. Most have specific structural requirements
+that are validated by the service, so conveying the format to the user ahead of
+time is critical to their experience.
+
+#### Why document value normalizations?
+
+While semantically a raw value is equal to its normalized form, they do not
+*appear* equal when using simple primitive comparison. As such, consumers need
+to know how to normalize a formatted field value in order to properly compare them.
```

Comment:
Well, normalization is one form of comparison - it's not the only way of performing an equivalence comparison.
I agree with @loudej that we should be emphasizing equivalence - but I'm not sure how we do that without going down a slippery slope.

Reply:
Yep, I've removed this language now as I've discouraged primitive text comparison and encouraged leveraging RFC compliant implementations for equality checks. PTAL.